iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/metadata/fetcher/cache/TableDeviceCacheEntry.java

```diff
@@ -36,32 +46,144 @@ public class TableDeviceCacheEntry {
+    if (Objects.isNull(map)) {
+      return 0;
+    }
+    final AtomicInteger diff = new AtomicInteger(0);
     updateMap.forEach(
         (k, v) -> {
           if (Objects.nonNull(v)) {
-            attributeMap.put(k, v);
+            if (!map.containsKey(k)) {
+              k = DataNodeTableCache.getInstance().tryGetInternColumnName(database, tableName, k);
+              diff.addAndGet(RamUsageEstimator.NUM_BYTES_OBJECT_REF);
+            }
+            diff.addAndGet(
+                (int) (RamUsageEstimator.sizeOf(v) - RamUsageEstimator.sizeOf(map.put(k, v))));
           } else {
-            attributeMap.remove(k);
+            map.remove(k);
+            diff.addAndGet((int) (-RamUsageEstimator.sizeOf(k) - RamUsageEstimator.sizeOf(v)));
+          }
+        });
+    // Typically the "update" and "invalidate" won't be concurrently called
+    // Here we reserve the check for consistency and potential safety
+    return Objects.nonNull(attributeMap.get()) ? diff.get() : 0;
+  }
+
+  public int invalidateAttribute() {
+    final AtomicInteger size = new AtomicInteger(0);
+    attributeMap.updateAndGet(
+        map -> {
+          if (Objects.nonNull(map)) {
+            size.set(
+                (int)
+                    (RamUsageEstimator.NUM_BYTES_OBJECT_REF * map.size()
+                        + map.values().stream()
+                            .mapToLong(RamUsageEstimator::sizeOf)
+                            .reduce(0, Long::sum)));
           }
+          return null;
         });
+    return size.get();
   }
 
   public String getAttribute(final String key) {
-    return attributeMap.get(key);
+    final Map<String, String> map = attributeMap.get();
+    return Objects.nonNull(map) ? map.get(key) : null;
   }
 
   public Map<String, String> getAttributeMap() {
-    return attributeMap;
+    return attributeMap.get();
+  }
+
+  /////////////////////////////// Last Cache ///////////////////////////////
+
+  public int updateLastCache(
+      final String database,
+      final String tableName,
+      final Map<String, TimeValuePair> measurementUpdateMap) {
+    return (lastCache.compareAndSet(null, new TableDeviceLastCache())
+            ? TableDeviceLastCache.INSTANCE_SIZE
+            : 0)
+        + tryUpdate(database, tableName, measurementUpdateMap);
+  }
+
+  public int tryUpdate(
+      final String database,
+      final String tableName,
+      final Map<String, TimeValuePair> measurementUpdateMap) {
+    final TableDeviceLastCache cache = lastCache.get();
+    final int result =
+        Objects.nonNull(cache) ? cache.update(database, tableName, measurementUpdateMap) : 0;
+    return Objects.nonNull(lastCache.get()) ? result : 0;
+  }
+
+  public TimeValuePair getTimeValuePair(final String measurement) {
+    final TableDeviceLastCache cache = lastCache.get();
+    return Objects.nonNull(cache) ? cache.getTimeValuePair(measurement) : null;
+  }
+
+  // Shall pass in "null" if last by time
```

Comment:
empty string for time column which is consistent with tsfile for aligned timeseries's time column